### PR TITLE
Fix #invariant Assertion in CacheFile

### PR DESCRIPTION
**Pull request by original-brownbear**
This assertion can trip pretty easily now that we made everything non-blocking. The ref count is decreased *before* the file is physically deleted so 

```java
assert evicted.get() == false || refCounter.refCount() != 0 || Files.notExists(file);
```
in `invariant` can trip when the assertion is checked before the delete on another thread has gone through but after it finished counting down the ref counter to `0`.

Not 100% sure about this one. It's a small performance hit when closing index inputs I guess (seems practically irrelevant though) but also it seems weird to add another layer of sync just to fix an assertion (otoh we already synchronize on file opening so maybe it's all good and it's nice to have the assertion I suppose).

Obvious alternative to this PR is to just drop `invariant` and maybe just deal with asserting that the cache is actually cleared in the end by enhancing tests to check for this in an `@After` hook or so

### CODE REVIEW

**Comment by tlrx** [comment]
Maybe we could remove the assertion in `invariant()` and instead assert that the `CacheFile` is indeed evicted at the time it is deleted from disk + assert that it is also evicted and the file deleted after the instance of `FileChannelReference` that triggers the closing of the `refCounter` is closed?

**Comment by original-brownbear** [comment]
Thanks for taking a look @tlrx 

>  (we'll need a specific implementation of RefCounted for this)

Yea this changes the mechanics quite a bit. I wonder if this is actually worth it, since in the end it seems that would require about the same amount of extra logging and even more code complexity than what I did here?

**Review by tlrx — APPROVED** [review]
LGTM, thanks for the extra iterations :+1:

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by original-brownbear** [comment]
Jenkins run  elasticsearch-ci/1(known master failover health check test failure)

**Comment by original-brownbear** [comment]
Thanks Tanguy!

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 1]
```diff
@@ -191,6 +192,11 @@ private boolean assertNoPendingListeners() {
         return true;
     }
 
+    private void decrementRefCount() {
+        final boolean released = refCounter.decRef();
+        assert released == false || Files.notExists(file);
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:197]: Can we also assert:

```
assert released == false || evicted.get();
```
?

Comment (original-brownbear) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:197]: Yea that sounds reasonable :) I'll open a PR

Comment (original-brownbear) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:197]: https://github.com/elastic/elasticsearch/pull/64272

**Pull request merged**